### PR TITLE
IPv6/multi : lowered the complexity of FreeRTOS_ARP.c

### PR DESCRIPTION
**Pull request by htibosch**
Description
-----------
In this PR I have split-up several functions in FreeRTOS_ARP.c
The complexity scores are now between 0 and 8, coming from :
~~~
Score | ln-ct | nc-lns| file-name(line): proc-name
   15     114      61   FreeRTOS_ARP.c(80): eARPProcessPacket
   37     155      94   FreeRTOS_ARP.c(239): vARPRefreshCacheEntry
total nc-lns      400
~~~
These are the new helper functions:
~~~
static eARPLookupResult_t prvCacheLookup( uint32_t ulAddressToLookup,
                                          MACAddress_t * const pxMACAddress,
                                          NetworkEndPoint_t ** ppxEndPoint );

static void vARPProcessPacketRequest( ARPPacket_t * pxARPFrame,
                                      NetworkEndPoint_t * pxTargetEndPoint,
                                      uint32_t ulSenderProtocolAddress );

static void vARPProcessPacketReply( ARPPacket_t * pxARPFrame,
                                    NetworkEndPoint_t * pxTargetEndPoint,
                                    uint32_t ulSenderProtocolAddress );

static eARPLookupResult_t eARPGetCacheEntryGateWay( uint32_t * pulIPAddress,
                                                    MACAddress_t * const pxMACAddress,
                                                    struct xNetworkEndPoint ** ppxEndPoint );

static void vARPSendRequestPacket( NetworkBufferDescriptor_t * pxNetworkBuffer );

static void prvFindCachEntry( const MACAddress_t * pxMACAddress,
                              const uint32_t ulIPAddress,
                              struct xNetworkEndPoint * pxEndPoint,
                              CacheLocation_t * pxLocation );
~~~

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by AniruddhaKanhere — CHANGES_REQUESTED** [review]

File: FreeRTOS_ARP.c [hunk 1]
```diff
@@ -254,6 +203,111 @@ eFrameProcessingResult_t eARPProcessPacket( NetworkBufferDescriptor_t * pxNetwor
...
+                                    uint32_t ulSenderProtocolAddress )
+{
+    ARPHeader_t * pxARPHeader = &( pxARPFrame->xARPHeader );
+
+    iptracePROCESSING_RECEIVED_ARP_REPLY( ulTargetProtocolAddress );
+    vARPRefreshCacheEntry( &( pxARPHeader->xSenderHardwareAddress ), ulSenderProtocolAddress, pxTargetEndPoint );
+    /* Process received ARP frame to see if there is a clash. */
+    #if ( ipconfigARP_USE_CLASH_DETECTION != 0 )
+        {
+            if( pxSourceEndPoint != NULL )
```
[the remark this answers is not in the document]
Comment (htibosch) [reply FreeRTOS_ARP.c:300]: Oops, I did not compile/test with `ipconfigARP_USE_CLASH_DETECTION` defined. Thanks for seeing this.

File: FreeRTOS_ARP.c [hunk 2]
```diff
@@ -324,141 +372,52 @@ void vARPRefreshCacheEntry( const MACAddress_t * pxMACAddress,
...
-                 * needs to add an entry that does not already exist. */
-                ucMinAgeFound = xARPCache[ x ].ucAge;
-                xUseEntry = x;
-            }
-            else
-            {
-                /* Nothing happens to this cache entry for now. */
-            }
-        }
+        prvFindCachEntry( pxMACAddress, ulIPAddress, pxEndPoint, &( xLocation ) );
```
[the remark this answers is not in the document]
Comment (htibosch) [reply FreeRTOS_ARP.c:380]: `s/prvFindCachEntry/prvFindCacheEntry/g`

File: include/FreeRTOS_ARP.h [hunk 3]
```diff
@@ -63,6 +63,15 @@
         eCantSendPacket    /**< 2 There is no IP address, or an ARP is still in progress, so the packet cannot be sent. */
     } eARPLookupResult_t;
 
+/** @brief A structure used internally in FreeRTOS_ARP.c.
+ * It is used as a parameter for the function prvFindCachEntry().*/
```
[the remark this answers is not in the document]
Comment (htibosch) [reply include/FreeRTOS_ARP.h:67]: Done

**Review by htibosch — COMMENTED** [review]
Thanks for reviewing, you had some good points.

**Comment by AniruddhaKanhere** [comment]
@htibosch, this comment needs to be addressed: "xAddressIsLocal is out of scope here. It was present in the vARPRefreshCacheEntry."

**Review by AniruddhaKanhere — APPROVED** [review]

**Review by cobusve — APPROVED** [review]

**Pull request merged**